core/src/main/java/org/apache/accumulo/core/clientImpl/ThriftTransportPool.java

```diff
@@ -690,4 +753,8 @@ private void shutdown() {
               + "all AccumuloClients being closed or garbage collected.");
     return cache;
   }
+
+  private Lock getLock(final ThriftTransportKey key) {
```

Comment:
I think the getLock() and getCache() methods could be replaced with a single method like the following

```java
  private LocksAndCache getLocksAndCache() {
    // very important to only read volatile once into a local variable as multiple reads may see different things.
    LocksAndCache local  = locksAndCache;
    if (local == null)
      throw new TransportPoolShutdownException(
          "The Accumulo singleton for connection pooling is disabled.  This is likely caused by "
              + "all AccumuloClients being closed or garbage collected.");
    return local;
  }
```
The code should be analyzed to ensure external operations only retrieve this once. Or if it is retrieved multiple times, analysis should be done to ensure the code handles changes caused by shutdown.  

Any code retrieving this should properly handle a concurrent shutdown.  Some code already does this.

Reply:
Made this changed, but the method is `getConnectionPool()` instead. I believe all calls to this are handling concurrent shutdown, with the exception of `returnTransport()` which I'm still not entirely sure about.

Reply:
I need to look at returnTransport() in the changes I made.

Reply:
`returnTransport()` LGTM.  I did make a comment elsewhere about moving access to the map outside of the lock.